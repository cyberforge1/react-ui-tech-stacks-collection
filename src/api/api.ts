// src/api/api.ts

import axios, { AxiosResponse } from 'axios';
import { Todo, TodoCreation, TodoUpdate, TodoListResponse, TodoResponse, ApiResponse } from '../types/types';

const API_BASE_URL = 'http://localhost:5001/api'; // Flask API base URL

// Get the main welcome message
export const getMainMessage = async (): Promise<string> => {
  try {
    const response: AxiosResponse<ApiResponse> = await axios.get(`${API_BASE_URL}/`);
    return response.data.message;
  } catch (error) {
    console.error('Error fetching Main message:', error);
    throw error;
  }
};

// Get Hello World message
export const getHelloWorld = async (): Promise<string> => {
  try {
    const response: AxiosResponse<ApiResponse> = await axios.get(`${API_BASE_URL}/helloworld/`);
    return response.data.message;
  } catch (error) {
    console.error('Error fetching Hello World message:', error);
    throw error;
  }
};

// Get all todos
export const getTodos = async (): Promise<Todo[]> => {
  try {
    const response: AxiosResponse<TodoListResponse> = await axios.get(`${API_BASE_URL}/todos/`);
    return response.data.todos;
  } catch (error) {
    console.error('Error fetching todos:', error);
    throw error;
  }
};

// Get a single todo by ID
export const getTodoById = async (id: number): Promise<Todo> => {
  try {
    const response: AxiosResponse<TodoResponse> = await axios.get(`${API_BASE_URL}/todos/${id}/`);
    return response.data.todo!;
  } catch (error) {
    console.error(`Error fetching todo with ID ${id}:`, error);
    throw error;
  }
};

// Create a new todo
export const createTodo = async (todo: TodoCreation): Promise<Todo> => {
  try {
    const response: AxiosResponse<TodoResponse> = await axios.post(`${API_BASE_URL}/todos/`, todo);
    return response.data.todo!;
  } catch (error) {
    console.error('Error creating todo:', error);
    throw error;
  }
};

// Update an existing todo
export const updateTodo = async (id: number, todoUpdate: TodoUpdate): Promise<Todo> => {
  try {
    const response: AxiosResponse<TodoResponse> = await axios.put(`${API_BASE_URL}/todos/${id}/`, todoUpdate);
    return response.data.todo!;
  } catch (error) {
    console.error(`Error updating todo with ID ${id}:`, error);
    throw error;
  }
};

// Delete a todo
export const deleteTodo = async (id: number): Promise<ApiResponse> => {
  try {
    const response: AxiosResponse<ApiResponse> = await axios.delete(`${API_BASE_URL}/todos/${id}/`);
    return response.data;
  } catch (error) {
    console.error(`Error deleting todo with ID ${id}:`, error);
    throw error;
  }
};
