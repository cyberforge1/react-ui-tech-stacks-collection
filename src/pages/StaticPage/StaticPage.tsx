// src/pages/StaticPage/StaticPage.tsx

import React, { useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { getMainMessage } from '../../api/api';

const StaticPage: React.FC = () => {
  const navigate = useNavigate();
  const [message, setMessage] = useState<string>('Welcome');
  const [error, setError] = useState<string | null>(null);
  const [isLoading, setIsLoading] = useState<boolean>(true);

  useEffect(() => {
    const fetchMessage = async () => {
      try {
        const fetchedMessage = await getMainMessage();
        setMessage(fetchedMessage);
        setError(null);
      } catch {
        setError('Failed to load the message. Please try again later.');
      } finally {
        setIsLoading(false);
      }
    };

    fetchMessage();
  }, []);

  const redirectToForm = () => {
    navigate('/form');
  };

  return (
    <div data-testid="static-page">
      {isLoading ? (
        <p>Loading...</p>
      ) : error ? (
        <p style={{ color: 'red' }}>{error}</p>
      ) : (
        <h1>{message}</h1>
      )}
      <button onClick={redirectToForm} aria-label="Go to Form">
        Go to Form
      </button>
    </div>
  );
};

export default StaticPage;
