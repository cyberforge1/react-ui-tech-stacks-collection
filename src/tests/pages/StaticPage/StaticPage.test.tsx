// src/tests/pages/StaticPage/StaticPage.test.tsx

import { render, screen, cleanup, fireEvent } from '@testing-library/react';
import { describe, it, expect, afterEach, beforeEach, vi } from 'vitest'; // Added beforeEach import
import StaticPage from '../../../pages/StaticPage/StaticPage';

// Mock the API calls used in StaticPage
vi.mock('../../../api/api', () => ({
  getMainMessage: vi.fn(),
}));

// Mock the useNavigate hook from react-router-dom
vi.mock('react-router-dom', async () => {
  const actual = await vi.importActual('react-router-dom');
  return {
    ...actual,
    useNavigate: vi.fn(), // Mock useNavigate as a vi.fn() function
  };
});

import { getMainMessage } from '../../../api/api';
import { useNavigate } from 'react-router-dom';

describe('StaticPage Component', () => {
  const navigateMock = vi.fn(); // Define the navigateMock

  beforeEach(() => {
    (useNavigate as ReturnType<typeof vi.fn>).mockReturnValue(navigateMock);
  });

  afterEach(() => {
    cleanup();
    vi.clearAllMocks(); // Clear all mocks after each test
  });

  it('displays "Loading..." initially', () => {
    render(<StaticPage />);
    const loadingText = screen.getByText(/loading\.\.\./i);
    expect(loadingText).toBeTruthy();
  });

  it('renders the fetched message from the API', async () => {
    (getMainMessage as ReturnType<typeof vi.fn>).mockResolvedValueOnce('Hello, World!');

    render(<StaticPage />);
    expect(await screen.findByText(/hello, world!/i)).toBeTruthy();
  });

  it('displays an error message when the API call fails', async () => {
    (getMainMessage as ReturnType<typeof vi.fn>).mockImplementationOnce(() =>
      Promise.reject(new Error('Failed to fetch message'))
    );

    render(<StaticPage />);
    expect(await screen.findByText(/failed to load the message\. please try again later\./i)).toBeTruthy();
  });

  it('navigates to the Form page when the button is clicked', () => {
    render(<StaticPage />);
    const button = screen.getByRole('button', { name: /go to form/i });
    fireEvent.click(button);

    expect(navigateMock).toHaveBeenCalledWith('/form');
  });

  it('matches the snapshot', () => {
    const { container } = render(<StaticPage />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
