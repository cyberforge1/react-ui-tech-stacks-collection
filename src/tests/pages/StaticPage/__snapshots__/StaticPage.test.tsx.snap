// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StaticPage Component > matches the snapshot 1`] = `
<div
  data-testid="static-page"
>
  <p>
    Loading...
  </p>
  <button
    aria-label="Go to Form"
  >
    Go to Form
  </button>
</div>
`;
